src/tango/ip/fd_ip.c

```diff
@@ -0,0 +1,381 @@
+#include "fd_ip.h"
+
+#include <arpa/inet.h>
+
+ulong
+fd_ip_align( void ) {
+  return FD_IP_ALIGN;
+}
+
+
+ulong
+fd_ip_footprint( ulong arp_entries,
+                 ulong route_entries ) {
+
+  ulong l;
+
+  l = FD_LAYOUT_INIT;
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_ip_t)                             );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_nl_t)                             );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, arp_entries   * sizeof(fd_nl_arp_entry_t)   );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, route_entries * sizeof(fd_nl_route_entry_t) );
+
+  return FD_LAYOUT_FINI( l, FD_IP_ALIGN );
+}
+
+
+void *
+fd_ip_new( void * shmem,
+           ulong  arp_entries,
+           ulong  route_entries ) {
+  if( !fd_ulong_is_aligned( (ulong)shmem, FD_IP_ALIGN ) ) {
+    FD_LOG_ERR(( "Attempt to fd_ip_new with unaligned memory" ));
+    return NULL;
+  }
+
+  ulong l;
+  uchar * mem = (uchar*)shmem;
+
+  l = FD_LAYOUT_INIT;
+
+  fd_ip_t * ip = (fd_ip_t*)mem;
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_ip_t)                             );
+
+  ulong ofs_netlink = l;
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, sizeof(fd_nl_t)                             );
+
+  ulong ofs_arp_table   = FD_ULONG_ALIGN_UP( l, FD_IP_ALIGN );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, arp_entries   * sizeof(fd_nl_arp_entry_t)   );
+
+  ulong ofs_route_table = FD_ULONG_ALIGN_UP( l, FD_IP_ALIGN );
+  l = FD_LAYOUT_APPEND( l, FD_IP_ALIGN, route_entries * sizeof(fd_nl_route_entry_t) );
+
+  ulong mem_sz = FD_LAYOUT_FINI( l, FD_IP_ALIGN );
+
+  /* clear all to zero */
+  fd_memset( ip, 0, mem_sz );
+
+  /* set values in ip */
+  ip->magic                = FD_IP_MAGIC;
+  ip->num_arp_entries      = arp_entries;
+  ip->num_route_entries    = route_entries;
+  ip->ofs_netlink          = ofs_netlink;
+  ip->ofs_arp_table        = ofs_arp_table;
+  ip->ofs_route_table      = ofs_route_table;
+
+  return (void*)ip;
+}
+
+
+fd_ip_t *
+fd_ip_join( void * mem ) {
+  if( !fd_ulong_is_aligned( (ulong)mem, FD_IP_ALIGN ) ) {
+    FD_LOG_ERR(( "Attempt to fd_ip_join with unaligned memory" ));
+    return NULL;
+  }
+
+  fd_ip_t * ip = (fd_ip_t*)mem;
+
+  if( ip->magic != FD_IP_MAGIC ) {
+    FD_LOG_ERR(( "Failed to fd_ip_join. Possibly memory corrupt" ));
+    return NULL;
+  }
+
+  /* initialize netlink */
+  fd_nl_t * netlink = fd_ip_netlink_get( ip );
+  if( fd_nl_init( netlink, 0 ) ) {
+    FD_LOG_ERR(( "Failed to initialize fd_netlink." ));
+    return NULL;
+  }
+
+  return ip;
+}
+
+
+void
```

Comment:
Can you align all these semantics to match the existing impls. Eg cnc / mcache / fctl, ...

leave returns a `void *` and will log warning for NULL ip etc. There's a few inconsistencies like this throughout this interface.